### DISCUSSION TITLE
第三课的关于ETH转USD，单位换算的问题

### DISCUSSION THREAD

**Discussion by henry652341974** [smartcontractkit/Web3_tutorial_Chinese#6] [Q&A] [answered]
要求执行合约支付的金额必须大于1USD
我是通过chainlink的datafee来实现换算
具体源码如下：
`
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/shared/interfaces/AggregatorV3Interface.sol";
contract FundMe {
uint256 MINIMUM_VALUE = 1 * 10 ** 18;
AggregatorV3Interface internal dataFeed =
AggregatorV3Interface(0x694AA1769357215DE4FAC081bf1f309aDC325306);

function fund() external payable {
// VALUE如果直接付的是1ETH,msg.value=1*10**18WEI
// msg.value的值都是以wei为单位
require(convertEhtToUsd(msg.value) >= MINIMUM_VALUE, "Send more eth");
}

// chainlink的接口获取ETH兑USD的价格，获得的结果是262438000000
function getChainlinkDataFeedLatestAnswer() public view returns (int) {
// prettier-ignore
(
/* uint80 roundID */,
int answer,
/uint startedAt/,
/uint timeStamp/,
/uint80 answeredInRound/
) = dataFeed.latestRoundData();
return answer;
}

// 用来转换Eth和Usd的方法，传入的参数是用户输入的value这里的值
// 是wei为单位
function convertEhtToUsd (uint256 ethAmount) internal view returns (uint256) {
uint256 ethPrice = uint256(getChainlinkDataFeedLatestAnswer());
return ethAmount * ethPrice / (108);
}
}
`
问题是：convertEhtToUsd方法这里，我理解的是getChainlinkDataFeedLatestAnswer方法得到的是一个262438000000，精度是108，实际1Eth=2624.38Usd,所以按照上面代码最后行"ethAmount * ethPrice / (108);"的意思是：ethPrice / (108)应该得到的是1wei usd的单价，因为传入参数的ethPrice的单位是wei，也就是1wei=2624.38usd。但这跟实际的1eth=2624.38usd明细是不一致的，因为1eth和1wei的换算是1eth=1*10**18wei。在线求解~！谢谢啊

**Comment by QingyangKong** [comment]
ethPrice 是 10**8
ethAmount 是 10**18
与之比较的 MINIMUM_VALUE 是 10**18
所以要把 ethPrice 的精度去掉，这一步通过 ethPrice * ethAmount / 10**8 实现

**Reply by henry652341974** [reply]
![image](https://github.com/user-attachments/assets/5362d115-14c4-41c7-9039-f0fb99f4b140)
老师的代码应该是写错了，麻烦看下

**Reply by henry652341974** [reply]
![image](https://github.com/user-attachments/assets/bf2187fc-4ff1-4566-9045-a91d5f60ff18)
这里也应该是写错了

**Comment by henry652341974** [comment, marked as answer]
我搞 明白了，你把下面该除10的18次方，扔给了MINIMUN_VALUE让他乘10的18次方，所以最终结果还是对的。但老师，你这样讲学生会很懵逼的

**Reply by QingyangKong** [reply]
嗯，主要是因为在 Solidity 中没有浮点数，一般会把精度体现在变量中。

**Reply by sub-bucket** [reply]
@QingyangKong 你好，这个ethPrice / (10 8) 是1个eth对应的usd的精度，那么这个108是怎么规定的呢，怎么来的呢，为什么又要除以 10 ** 8，是因为datafeed的时候乘以了10**8的原因么，还是为什么

**Reply by QingyangKong** [reply]
对 USD 的精度是 8，对 ETH 的精度是 18，这个是喂价的精度标准
